### PR TITLE
[new release] mirage-block-unix (2.13.0)

### PR DESCRIPTION
**Pull request by dinosaure**
MirageOS disk block driver for Unix

- Project page: <a href="https://github.com/mirage/mirage-block-unix">https://github.com/mirage/mirage-block-unix</a>
- Documentation: <a href="https://mirage.github.io/mirage-block-unix/">https://mirage.github.io/mirage-block-unix/</a>

##### CHANGES:

* Add a configuration field for the size of sectors (mirage/mirage-block-unix#106, @Julow, @hannesm, @dinosaure)
* Update to cstruct.6.0.0 (mirage/mirage-block-unix#108, @MisterDA, @dinosaure)
* Depends on `io-page.unix` instead of `io-page-unix` (@dinosaure)

### CODE REVIEW

**Comment by mseri** [comment]
Are these due to breaking changes from this package?
```
#=== ERROR while compiling fat-filesystem.0.15.0 ==============================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.13.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.13/.opam-switch/build/fat-filesystem.0.15.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p fat-filesystem -j 31
# exit-code            1
# env-file             ~/.opam/log/fat-filesystem-8089-60459a.env
# output-file          ~/.opam/log/fat-filesystem-8089-60459a.out
### output ###
[...]
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlc.opt -w -40 -g -bin-annot -I src/.fat_filesystem.objs/byte -I /home/opam/.opam/4.13/lib/astring -I /home/opam/.opam/4.13/lib/bigarray-compat -I /home/opam/.opam/4.13/lib/bytes -I /home/opam/.opam/4.13/lib/cstruct -I /home/opam/.opam/4.13/lib/fmt -I /home/opam/.opam/4.13/lib/io-page -I /home/opam/.opam/4.13/lib/lwt -I /home/opam/.opam/4.13/lib/mirage-block -I /home/opam/.opam/4.13/lib/mirage-kv -I /home/opam/.opam/4.13/lib/re -I /home/opam/.opam/4.13/lib/re/str -I /home/opam/.opam/4.13/lib/result -I /home/opam/.opam/4.13/lib/rresult -I /home/opam/.opam/4.13/lib/seq -I /home/opam/.opam/4.13/lib/stdlib-shims -no-alias-deps -o src/.fat_filesystem.objs/byte/fat.cmi -c -intf src/fat.pp.mli)
# File "src/fat.mli", line 51, characters 10-25:
# 51 |   include Mirage_device.S
#                ^^^^^^^^^^^^^^^
# Error: Unbound module Mirage_device
```

```
#=== ERROR while compiling qcow.0.11.0 ========================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.13.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.13/.opam-switch/build/qcow.0.11.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p qcow -j 31
# exit-code            1
# env-file             ~/.opam/log/qcow-51-80d8de.env
# output-file          ~/.opam/log/qcow-51-80d8de.out
### output ###
[...]
# Error: This pattern matches values of type [? `Unimplemented ]
#        but a pattern was expected which matches values of type B.error
#        The second variant type is private,
#        it may not allow the tag(s) `Unimplemented
#     ocamlopt lib/.qcow.objs/native/qcow_cluster_map.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -w -40 -g -I lib/.qcow.objs/byte -I lib/.qcow.objs/native -I /home/opam/.opam/4.13/lib/angstrom -I /home/opam/.opam/4.13/lib/asetmap -I /home/opam/.opam/4.13/lib/astring -I /home/opam/.opam/4.13/lib/base/caml -I /home/opam/.opam/4.13/lib/bigarray-compat -I /home/opam/.opam/4.13/lib/bigstringaf -I /home/opam/.opam/4.13/lib/bytes -I /home/opam/.opam/4.13/lib/cstruct -I /home/opam/.opam/4.13/lib/cstruct-lwt -I /home/opam/.opam/4.13/lib/fmt -I /home/opam/.opam/4.13/lib/io-page -I /home/opam/.opam/4.13/lib/io-page-unix -I /home/opam/.opam/4.13/lib/logs -I /home/opam/.opam/4.13/lib/lwt -I /home/opam/.opam/4.13/lib/lwt/unix -I /home/opam/.opam/4.13/lib/mirage-block -I /home/opam/.opam/4.13/lib/mirage-block-unix -I /home/opam/.opam/4.13/lib/mirage-time -I /home/opam/.opam/4.13/lib/mirage-types -I /home/opam/.opam/4.13/lib/mmap -I /home/opam/.opam/4.13/lib/ocaml/threads -I /home/opam/.opam/4.13/lib/ocplib-endian -I /home/opam/.opam/4.13/lib/ocplib-endian/bigstring -I /home/opam/.opam/4.13/lib/parsexp -I /home/opam/.opam/4.13/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.13/lib/prometheus -I /home/opam/.opam/4.13/lib/re -I /home/opam/.opam/4.13/lib/result -I /home/opam/.opam/4.13/lib/rresult -I /home/opam/.opam/4.13/lib/seq -I /home/opam/.opam/4.13/lib/sexplib -I /home/opam/.opam/4.13/lib/sexplib0 -I /home/opam/.opam/4.13/lib/stdlib-shims -I /home/opam/.opam/4.13/lib/stringext -I /home/opam/.opam/4.13/lib/uri -intf-suffix .ml -no-alias-deps -o lib/.qcow.objs/native/qcow_cluster_map.cmx -c -impl lib/qcow_cluster_map.pp.ml)
# File "lib/qcow_cluster_map.ml", line 345, characters 28-61:
# 345 |     ~read_cluster:(fun _ -> Lwt.return (Error `Unimplemented))
#                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This expression has type
#          (Cstruct.t, [> `Unimplemented ] as 'a) result Lwt.t
#        but an expression was expected of type
#          (Cstruct.t, Mirage_block.error) result Lwt.t
#        Type [> `Unimplemented ] as 'a is not compatible with type
#          Mirage_block.error = [ `Disconnected ] 
#        The second variant type does not allow tag(s) `Unimplemented
#     ocamlopt lib/.qcow.objs/native/qcow_padded.{cmx,o} (exit 2)
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -w -40 -g -I lib/.qcow.objs/byte -I lib/.qcow.objs/native -I /home/opam/.opam/4.13/lib/angstrom -I /home/opam/.opam/4.13/lib/asetmap -I /home/opam/.opam/4.13/lib/astring -I /home/opam/.opam/4.13/lib/base/caml -I /home/opam/.opam/4.13/lib/bigarray-compat -I /home/opam/.opam/4.13/lib/bigstringaf -I /home/opam/.opam/4.13/lib/bytes -I /home/opam/.opam/4.13/lib/cstruct -I /home/opam/.opam/4.13/lib/cstruct-lwt -I /home/opam/.opam/4.13/lib/fmt -I /home/opam/.opam/4.13/lib/io-page -I /home/opam/.opam/4.13/lib/io-page-unix -I /home/opam/.opam/4.13/lib/logs -I /home/opam/.opam/4.13/lib/lwt -I /home/opam/.opam/4.13/lib/lwt/unix -I /home/opam/.opam/4.13/lib/mirage-block -I /home/opam/.opam/4.13/lib/mirage-block-unix -I /home/opam/.opam/4.13/lib/mirage-time -I /home/opam/.opam/4.13/lib/mirage-types -I /home/opam/.opam/4.13/lib/mmap -I /home/opam/.opam/4.13/lib/ocaml/threads -I /home/opam/.opam/4.13/lib/ocplib-endian -I /home/opam/.opam/4.13/lib/ocplib-endian/bigstring -I /home/opam/.opam/4.13/lib/parsexp -I /home/opam/.opam/4.13/lib/ppx_sexp_conv/runtime-lib -I /home/opam/.opam/4.13/lib/prometheus -I /home/opam/.opam/4.13/lib/re -I /home/opam/.opam/4.13/lib/result -I /home/opam/.opam/4.13/lib/rresult -I /home/opam/.opam/4.13/lib/seq -I /home/opam/.opam/4.13/lib/sexplib -I /home/opam/.opam/4.13/lib/sexplib0 -I /home/opam/.opam/4.13/lib/stdlib-shims -I /home/opam/.opam/4.13/lib/stringext -I /home/opam/.opam/4.13/lib/uri -intf-suffix .ml -no-alias-deps -o lib/.qcow.objs/native/qcow_padded.cmx -c -impl lib/qcow_padded.pp.ml)
# File "lib/qcow_padded.ml", line 26, characters 73-74:
# 26 |     | e -> Format.kasprintf Lwt.fail_with "Unknown error: %a" B.pp_error e
#                                                                               ^
# Error: This expression has type [> `Disconnected | `Unimplemented ]
#        but an expression was expected of type B.error
#        The second variant type is private,
#        it may not allow the tag(s) `Unimplemented
```

**Comment by dinosaure** [comment]
I'm not sure, may be is related to an old release of `mirage-block`/`mirage-device`. I will try locally to see if they are installable.

**Comment by mseri** [comment]
There seem to be the same failures still

**Comment by hannesm** [comment]
We'll need upper bounds for qcow, fat-filesystem, shared-block-ring then. From the error log, it looks like the removal of mirage-device (in mirage-block 3.0.0) requires changes to the 3 packages above. I'll try to look into that later today. (mirage-block-unix 2.13 is the first release supporting mirage-block 3.0.0 (since https://github.com/mirage/mirage-block-unix/commit/aef82d3ffef469c507007423465ac67263a9b0e0)).

TL;DR: let's have some upper bounds for the failing packages above in opam-repository, and fix them upstream & cut releases.

**Comment by hannesm** [comment]
Further discussion / reasoning is at https://github.com/mirage/mirage/issues/1203

**Comment by dinosaure** [comment]
Oh ok, I was figure out it's possibly a `rresult` problem.

**Comment by kit-ty-kate** [comment]
Thanks!

**Comment by hannesm** [comment]
I opened PRs for fat-filesystem and shared-block-ring, but qcow is quite some work to get up to speed -- is anyone still using it /cc @djs55 (if yes, there are a bunch of PRs which could need some review / merge). If qcow is still in use, I can work later on bringing it up to speed with new APIs.
